### PR TITLE
Add justification and width arguments to silhouette plotting functions

### LINKED ISSUE #101: Add options for justification (e.g. vjust and hjust similar to geom_text())
**Is your feature request related to a problem? Please describe.**
Figure edge position is figure and size dependent, since they all centre on the center of the figure.

**Describe the solution you'd like**
It would be nice with an option like hjust and vjust (as in geom_text()), where vjust = .5 is default figure centre, and vjust = 0 means bottom edge of the figure is exactly on the provided y, and vjust = 1, means the top edge of the figure is exactly on the provided y. 

**Additional context**
An example figure here, where I wish to put all the phylopics bottom edges aligned with the bottom border of the plot:
ExtinctMammals_

Thanks for the package, and sorry if the option is available and I just couldn't find it.

### PR DESCRIPTION
**Pull request by willgearty**
This adds justification arguments/aesthetics to geom_phylopic, add_phylopic, and add_phylopic_base. They generally seem to work well, but there are some issues where the silhouette will over-justify due to interactions between the aspect ratio of the plot and the aspect ratio of the silhouette. I'll look into it a little more, but it might not be possible to address that 100%. It'd be great if you could test and review this @LewisAJones.

Fixes #101.

### CODE REVIEW

**Comment by pmur002** [comment]
Sounds like reasonable things to add/fix.  I will take a look.

**Comment by willgearty** [comment]
I've added width/height arguments (and deprecated size/ysize arguments), which fixes #103. For many cases, this actually seems to help the justification. For example, specifying `width` instead of `height` while changing the `hjust` appears to work well (and likewise with `height` while changing the `vjust`). However, there are still some issues where silhouettes will plot smaller than they should because of the interaction between the silhouette aspect ratio and the size and aspect ratio of the plot area. I believe that this should all be resolved if/when `grImport2::grid.picture` allows for the `width`/`height` to be NULL.

**Comment by willgearty** [comment]
@pmur002 any chance you've been able to look into this? Let me know if we can help in any way.

**Comment by pmur002** [comment]
Sorry, intense onset of teaching at the moment.  Feel free to keep prompting so that I do eventually get back to this.

**Comment by willgearty** [comment]
@pmur002 just sending another prompting your way to remind you about this.

**Comment by pmur002** [comment]
Thanks for the reminder!  Still snowed under with teaching, but it would help to clarify a couple of things:
are you still interested in hjust/vjust changes as well as width/height changes (in grImport2)?  can you please provide some as-simple-as-possible test cases that demonstrate the current problem?

**Comment by willgearty** [comment]
@pmur002 Ideally we'd like `grid.picture`/`pictureGrob` to function just like `grid.raster`/`rasterGrob`, including:

- Being able to have `width` or `height` be set to NULL and then autocalculated at render time (this functionality just appears to be missing from grImport2)
- Being able to specify `hjust`/`vjust` justification separately (currently `just` works, so this just seems like a minor bug where `hjust`/`vjust` aren't used if supplied)

So, to answer your question, yes, we are still interested in being able to have width/height and justification changes in grImport2. I think implementing the first point above should fix all of the problems in rphylopic (including the interaction between justification and size), but we'd appreciate the second point being fixed as well, if possible.

Here are some minimal test cases that demonstrate the current differences between the two sets of functions:

``` r
library(grid)
library(grImport2)
library(rphylopic)
uuid <- "16cfde1b-d577-4de8-82b9-62b760aacba5"
img_svg <- recolor_phylopic(get_phylopic(uuid, format = "vector"), fill = "blue")
img_png <- recolor_phylopic(get_phylopic(uuid, format = "raster"), fill = "red")

plot(1:10, type = "n")
# default width is NULL, which is autocalculated at render time
# horizontal justification works as expected
grid.raster(img_png, height = .25, x = .5, y = .5, just = c(0, 1))
# default width is unit(1, "npc")
# so horizontal justification doesn't work like in grid.raster
grid.picture(img_svg, height = .25, x = .5, y = .5, just = c(0, 1))
```



``` r

plot(1:10, type = "n")
# same problem with height and vertical justification
grid.raster(img_png, width = .25, x = .5, y = .5, just = c(0.5, 0))
grid.picture(img_svg, width = .25, x = .5, y = .5, just = c(0.5, 0))
```



``` r

# so, I guess we need to specify both the height and the width...
svg_ar <- rphylopic:::aspect_ratio(img_svg)

plot(1:10, type = "n")
grid.raster(img_png, width = .5, x = .5, y = .5, just = c(0.5, 0))
# except the plot window aspect ratio can change...
# doesn't match in size, but maybe matches in justification?:
grid.picture(img_svg, height = .5, width = .5 * svg_ar, x = .5, y = .5, just = c(0.5, 0))
```



``` r
# matches in size but not justification when plot window is taller:
```
image

<sup>Created on 2024-03-11 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by pmur002** [comment]
Awesome.  Thanks for the clarification and the examples.  It's on my list!

**Comment by willgearty** [comment]
@pmur002 any progress on this?

**Comment by pmur002** [comment]
Sorry, still behind on teaching.  Please keep reminding.

**Comment by pmur002** [comment]
The bad news:  no.  The good news:  getting closer.  A reminder in July or August might actually produce some action.  Hopefully.

**Comment by willgearty** [comment]
Hi @pmur002, I'm travelling for the next three weeks, so I wanted to give you a poke beforehand to see if you'll have time to work on this soon?

**Comment by willgearty** [comment]
@pmur002 ?

**Comment by pmur002** [comment]
It's a funny time of year for christmas, but I have (finally) committed an attempt at this request on [r-forge](https://r-forge.r-project.org/projects/grimport/).

Needs regression testing before it can go anywhere near CRAN, but you might like to take it for a spin to see if it works.

A couple of points about the examples:

1. To make a fair comparison with `grid.raster()` you need `grid.picture(..., expansion=0)`.  I think your third example actually works just with this.
2. To make a fair comparison, both `width` and `height` need to be the same in both `grid.raster()` and `grid.picture()` calls.  I think your third example (with taller device) has this complication.

Thanks heaps for your patience!  I hope this fix is along the right lines.

**Comment by pmur002** [comment]
{grImport2} version 0.3-3 is now on its way to CRAN

**Comment by willgearty** [comment]
Thank you so much for your changes to grImport2 @pmur002, everything looks good to me (and it's great to see it on CRAN already!). @LewisAJones I just need to add a few more tests then it should be ready for review.

**Comment by willgearty** [comment]
Phew, OK, I think that does it! @LewisAJones, it should be all set for you to take a look.

**Comment by pmur002** [comment]
> Thank you so much for your changes to grImport2 @pmur002, everything looks good to me (and it's great to see it on CRAN already!). 

No worries.  I think a turn around of 18 months qualifies as "the least I could do" :)

File: R/geom_phylopic.R [hunk 1]
```diff
@@ -400,20 +480,59 @@ phylopicGrob <- function(img, x, y, height, color, fill, alpha,
...
-                           default.units = "native")
+      img_grob <- rasterGrob(img, x = x, y = y,
+                             height = height, width = width,
+                             default.units = "native",
+                             hjust = hjust, vjust = vjust)
   }
   return(img_grob)
 }
 
+#' Scales for phylopic heighgt or width
```
Comment (LewisAJones) [inline R/geom_phylopic.R:502]:
```suggestion
#' Scales for phylopic height or width
```

File: tests/testthat/_snaps/geom_phylopic/phylopic-key-glyph-with-img.svg
Comment (LewisAJones) [inline tests/testthat/_snaps/geom_phylopic/phylopic-key-glyph-with-img.svg:1]: Should we be using two images (cat and walrus) for the glyph?

**Review by LewisAJones — CHANGES_REQUESTED** [review]
Hi @willgearty, I've been through this PR now and it is looking really good to me! I noticed a few things we might want to address/discuss but other than those very minor comments, the code and docs looks good to me and it seems to be working really well.

File: R/add_phylopic.r [hunk 2]
```diff
@@ -33,6 +42,16 @@
 #' @param vertical \code{logical}. Should the silhouette be flipped vertically?
 #' @param angle \code{numeric}. The number of degrees to rotate the silhouette
 #'   clockwise. The default is no rotation.
+#' @param hjust \code{numeric}. A numeric vector between 0 and 1 specifying
+#'   horizontal justification (left = 0, center = 0.5, right = 1). Note that,
```
[the remark this answers is not in the document]
Comment (willgearty) [reply R/add_phylopic.r:46]: Looks like this is actually a problem in clarity in the ggplot documentation. A value of 0 represents left-justification, which moves the item to the right, whereas a value of 1 represents right-justification which moves the item to the left.

[the remark this answers is not in the document]
Comment (willgearty) [reply R/add_phylopic.r:46]: See here: https://ggplot2.tidyverse.org/articles/ggplot2-specs.html#justification

File: R/add_phylopic.r [hunk 3]
```diff
@@ -33,6 +42,16 @@
...
+#'   due to the enforcement of the silhouette's aspect ratio, there may be
+#'   unexpected behavior due to interactions between the aspect ratio of the
+#'   plot and the aspect ratio of the silhouette.
+#' @param vjust \code{numeric}. A numeric vector between 0 and 1 specifying
+#'   vertical justification (top = 1, middle = 0.5, bottom = 0). Note that, due
```
[the remark this answers is not in the document]
Comment (willgearty) [reply R/add_phylopic.r:51]: I think there's some confusion around what "justification" means here. "top"-justification means that the "top" of the phylopic is aligned to the x,y coordinate (where the phylopic moves down relative to the default location), whereas "bottom"-justification means the "bottom" of the phylopic is aligned to x,y coordinate (where the phylopic moves up relative to the default location).

File: R/add_phylopic_base.r [hunk 4]
```diff
@@ -13,12 +13,23 @@
 #'   for images which allows commercial usage, and "sa" for images without a
 #'   ShareAlike clause. The user can also combine these filters as a vector.
 #' @param x \code{numeric}. x value of the silhouette center. Ignored if `y` and
-#'   `ysize` are not specified.
+#'   `ysize` are not specified. If "NULL", the default, the mean value of the
```
[the remark this answers is not in the document]
Comment (willgearty) [reply R/add_phylopic_base.r:16]: Ugh...

File: R/add_phylopic.r (hunk 2 above)
[the remark this answers is not in the document]
Comment (LewisAJones) [reply R/add_phylopic.r:46]: Well obviously we should stick with that if that is the standard, and frankly who am I to argue with the ggplot2 documentation, but it does seems a little counter-intuitive (I guess I should use ggplot2 more!) - I've always used justification and alignment interchangeably which is perhaps not really accurate. Thanks for clarifying!

File: tests/testthat/_snaps/geom_phylopic/phylopic-key-glyph-with-img.svg
Comment (willgearty) [reply tests/testthat/_snaps/geom_phylopic/phylopic-key-glyph-with-img.svg:1]: I suppose one could argue that this is also testing that a single image is reused properly if there are fewer images than there are legend items to show. But I think I was mostly just being lazy when I made the test 🙈 .

**Pull request merged**